File: typescript/day13_part1_2020.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').trim().split('\n');
const earliestTimestamp = Number(input[0]);
const busIDs = input[1].split(',').filter(id => id !== 'x').map(Number);

let earliestBusID = 0;
let minWaitTime = Infinity;

for (const busID of busIDs) {
    const waitTime = busID - (earliestTimestamp % busID);
    if (waitTime < minWaitTime) {
        minWaitTime = waitTime;
        earliestBusID = busID;
    }
}

console.log(earliestBusID * minWaitTime);

File: fsharp/day13_part1_2020.fs

module Day13

let input = System.IO.File.ReadAllLines "input.txt"
let earliestTimestamp = int input.[0]
let busIDs = input.[1].Split(',') |> Array.filter (fun x -> x <> "x") |> Array.map int

let findEarliestBus () =
    let mutable minWaitTime = System.Int32.MaxValue
    let mutable earliestBusID = 0
    for id in busIDs do
        let waitTime = id - (earliestTimestamp % id)
        if waitTime < minWaitTime then
            minWaitTime <- waitTime
            earliestBusID <- id
    minWaitTime * earliestBusID

printfn "%d" (findEarliestBus ())
